RIOT-OS/RIOT
# sys/net/nanocoap: fix `coap_build_reply_header()`

### Contribution description

This fixes the return value of `coap_build_reply_header()` to also be correct when the request used an extended length token (RFC 8974). In `master`, the extend token length field is not taken into account.

The nanocoap unit tests were extended to catch the bug.

### Testing procedure

Cherry-pick the second commit onto `master` and run:

```
 make BOARD=native -C tests/unittests all -j && make BOARD=native -C tests/unittests test
```

This will fail:

```
main(): This is RIOT! (Version: 2024.10-devel-373-g5ae4c)
Help: Press s to start test, r to print it is ready
READY
s
START
.......................==66399==WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!
.....................................................................................................................................................................................................................==66399==WARNING: ASan is ignoring requested __asan_handle_no_return: stack type: default top: 0xffb15000; bottom 0xea802000; size: 0x15313000 (355545088)
False positive error reports may follow
For details see https://github.com/google/sanitizers/issues/189
.................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
nanocoap_tests.test_nanocoap__token_length_ext_16 (tests/unittests/tests-nanocoap/tests-nanocoap.c 1126) exp 21 was 20
.
nanocoap_tests.test_nanocoap__token_length_ext_269 (tests/unittests/tests-nanocoap/tests-nanocoap.c 1176) exp 275 was 273
...........................................................................................................................................................................................................................................................................................................................................................
run 1209 failures 2
```

Run the same test again in this PR, this passes:

```
main(): This is RIOT! (Version: 2024.10-devel-375-gad7e45-sys/net/nanocoap/fix-coap_build_reply_header)
Help: Press s to start test, r to print it is ready
READY
s
START
.......................==69252==WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!
.....................................................................................................................................................................................................................==69252==WARNING: ASan is ignoring requested __asan_handle_no_return: stack type: default top: 0xffa32000; bottom 0xee502000; size: 0x11530000 (290652160)
False positive error reports may follow
For details see https://github.com/google/sanitizers/issues/189
.............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
OK (1209 tests)
```

### Issues/PRs references

None
